File: day1/solution.js
// It's not 5
// It's not -17

const { readFileSync } = require('fs');

const input = readFileSync('input.txt', 'utf-8')
  .split('\n')
  .map((x) => parseInt(x, 10))
  .filter((x) => !isNaN(x));

// Part 1
const part1 = input.reduce(((x, y) => x + y), 0);
console.log("part 1 is: " + part1);

// Part 2
// Recursive attempt
function firstDupe(seen, total, remaining, original) {
  console.log(`next is ${remaining[0]}, remaining ${remaining.length}, total is ${total}, seen ${seen.size}`)
  if (remaining.length === 0) {
    return firstDupe(seen, total, original, original)
  }

  total += remaining[0];
  if (seen.has(total)) {
    return total;
  }
  seen.add(total);
  return firstDupe(seen, total, remaining.slice(1), original);
}

// Iterative attempt
function repeatedTotal(changes) {
  let total = 0;
  let seen = new Set([0]);
  while(true) {
    for (let change of changes) {
      total += change;
      console.log(`next is ${change}, total is ${total}, seen ${seen.size}`)
      if (seen.has(total)) {
        return total;
      }
      seen.add(total);
    }
  }
}

console.log(repeatedTotal(input))
//const part2 = firstDupe(new Set([0]), 0, input, input)
//console.log("part 2 is: " + part2);

File: day1/hs/solution.hs
import System.IO
import qualified Data.Set as Set

-- Part 1 stuff

finalFreq :: Int -> [Int] -> Int
finalFreq acc [] = acc
finalFreq acc (change:changes) = finalFreq (acc + change) changes


-- Part 2 stuff

-- Either found a frequency, or need to continue with the specified state
data MaybeFreq = Found Int | Continue (Set.Set Int) Int

repeatHelper :: Set.Set Int -> Int -> [Int] -> MaybeFreq
repeatHelper seen currentFreq [] = Continue seen currentFreq
repeatHelper seen currentFreq (change:changes) =
  let newFreq = currentFreq + change
  in
    if newFreq `Set.member` seen then
      Found newFreq
    else
      repeatHelper (Set.insert newFreq seen) newFreq changes

findRepeat :: (Set.Set Int) -> Int -> [Int] -> Int
findRepeat seen current changes =
  case repeatHelper seen current changes of
    Found x -> x
    Continue seen' current' -> findRepeat seen' current' changes


-- Main program stuff

parseInt :: String -> Int
parseInt ('+':pos) = read pos
parseInt ('-':neg) = -1 * (read neg)

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  let changes = map parseInt (words contents)
      part1 = finalFreq 0 changes -- I guess probably reduce would have worked here. How does that work in haskell?
      part2 = findRepeat Set.empty 0 changes
  print part1
  print part2
  hClose handle
